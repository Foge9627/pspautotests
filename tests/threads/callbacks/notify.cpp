#include "shared.h"

#include <psppower.h>

int cbFunc(int arg1, int arg2, void *arg) {
	checkpoint(" * cbFunc hit: %08x, %08x, %08x", arg1, arg2, arg);
	return 0;
}

inline void testNotify(const char *title, SceUID cb, int arg) {
	int result = sceKernelNotifyCallback(cb, arg);
	if (result == 0) {
		checkpoint("%s: OK", title);
	} else {
		checkpoint("%s: Failed (%08x)", title, result);
	}
}

struct CallbackSleeper : public BasicThread {
	CallbackSleeper(const char *name, int prio = 0x60, bool doStart = true)
		: BasicThread(name, prio), ret_(0) {
		if (doStart) {
			start();
		}
	}

	static int callback(int arg1, int arg2, void *arg) {
		CallbackSleeper *me = (CallbackSleeper *)arg;
		return me->hit(arg1, arg2);
	}

	virtual int hit(int arg1, int arg2) {
		checkpoint("  Callback hit %08x, %08x, returning %08x", arg1, arg2, ret_);
		return ret_;
	}

	virtual int execute() {
		cb_ = sceKernelCreateCallback(name_, &CallbackSleeper::callback, (void *)this);
		checkpoint("  Beginning sleep on %s", name_);
		checkpoint("  Woke from sleep: %08x", sceKernelSleepThreadCB());
		return 0;
	}

	SceUID callbackID() {
		return cb_;
	}

	void wakeup() {
		sceKernelWakeupThread(thread_);
	}

	void setReturn(int ret) {
		ret_ = ret;
	}

	int ret_;
	SceUID cb_;
};

struct SelfNotifier : public CallbackSleeper {
	SelfNotifier(const char *name, int prio = 0x60)
		: CallbackSleeper(name, prio, false), done_(false) {
		start();
	}

	virtual int hit(int arg1, int arg2) {
		checkpoint("  Callback hit %08x, %08x, returning %08x", arg1, arg2, ret_);
		if (!done_) {
			done_ = true;
			checkpoint("  Notifying self within callback: %08x", sceKernelNotifyCallback(cb_, 0xABCD));
			checkpoint("  sceKernelDelayThreadCB within callback: %08x", sceKernelDelayThreadCB(1000));
		}
		return ret_;
	}

	bool done_;
};

extern "C" int main(int argc, char *argv[]) {
	SceUID cb = sceKernelCreateCallback("notify", &cbFunc, NULL);

	checkpointNext("Objects:");
	testNotify("  Normal", cb, 0x1337);
	testNotify("  NULL", 0, 0x1337);
	testNotify("  Invalid", 0xDEADBEEF, 0x1337);
	sceKernelDeleteCallback(cb);
	testNotify("  Deleted", cb, 0x1337);

	cb = sceKernelCreateCallback("notify", &cbFunc, NULL);
	checkpointNext("Values:");
	testNotify("  Zero", cb, 0);
	testNotify("  DEADBEEF", cb, 0xDEADBEEF);

	checkpointNext("Notifies:");
	int result = 0;
	for (int i = 0; i < 10000; ++i) {
		result = sceKernelNotifyCallback(cb, 1);
		if (result != 0) {
			checkpoint("  Failed at %d: %08x", i, result);
			break;
		}
	}
	if (result == 0) {
		checkpoint("  10000 notifies: OK");
	}

	checkpoint("sceKernelDelayThreadCB: %08x", sceKernelDelayThreadCB(1000));

	checkpointNext("Different thread:");
	{
		CallbackSleeper waiter1("better priority sleeping thread", 0x10);
		CallbackSleeper waiter2("worse priority sleeping thread", 0x30);
		sceKernelDelayThread(1000);
		sceKernelNotifyCallback(waiter1.callbackID(), 0x1337);
		sceKernelNotifyCallback(waiter2.callbackID(), 0x1337);
		sceKernelDelayThread(1000);
	}

	checkpointNext("Return value:");
	{
		CallbackSleeper waiter("sleeping thread");
		waiter.setReturn(0x1337);
		sceKernelDelayThread(1000);
		testNotify("  Notify #1", waiter.callbackID(), 0x1337);
		sceKernelDelayThread(1000);
		testNotify("  Notify #2", waiter.callbackID(), 0x1337);
		sceKernelDelayThread(1000);
	}

	checkpointNext("Recursion:");
	{
		SelfNotifier waiter("sleeping thread");
		sceKernelDelayThread(1000);
		testNotify("  Notify #1", waiter.callbackID(), 0x1337);
		sceKernelDelayThread(1000);
	}

	checkpointNext("Mixing types:");
	checkpoint("  scePowerRegisterCallback (causes notify): %08x", scePowerRegisterCallback(0, cb));
	testNotify("  Manual notify", cb, 0x1337);
	checkpoint("  sceKernelDelayThreadCB: %08x", sceKernelDelayThreadCB(1000));

	return 0;
}
