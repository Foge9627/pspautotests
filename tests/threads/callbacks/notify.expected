[x] Objects:
[x]   Normal: OK
[x]   NULL: Failed (800201a1)
[x]   Invalid: Failed (800201a1)
[x]   Deleted: Failed (800201a1)

[x] Values:
[x]   Zero: OK
[x]   DEADBEEF: OK

[x] Notifies:
[x]   10000 notifies: OK
[x]  * cbFunc hit: 00002712, 00000001, 00000000
[r] sceKernelDelayThreadCB: 00000000

[x] Different thread:
[x]   Beginning sleep on better priority sleeping thread
[x]   ** started better priority sleeping thread
[x]   ** started worse priority sleeping thread
[r]   Beginning sleep on worse priority sleeping thread
[r]   Callback hit 00000001, 00001337, returning 00000000
[r]   Callback hit 00000001, 00001337, returning 00000000
[r]   ** stopped worse priority sleeping thread
[r]   ** stopped better priority sleeping thread

[x] Return value:
[x]   ** started sleeping thread
[r]   Beginning sleep on sleeping thread
[r]   Notify #1: OK
[r]   Callback hit 00000001, 00001337, returning 00001337
[r]   Notify #2: Failed (800201a1)
[r]   ** stopped sleeping thread

[x] Recursion:
[x]   ** started sleeping thread
[r]   Beginning sleep on sleeping thread
[r]   Notify #1: OK
[r]   Callback hit 00000001, 00001337, returning 00000000
[r]   Notifying self within callback: 00000000
[r]   Callback hit 00000001, 0000abcd, returning 00000000
[r]   sceKernelDelayThreadCB within callback: 00000000
[r]   ** stopped sleeping thread

[x] Mixing types:
[x]   scePowerRegisterCallback (causes notify): 00000000
[x]   Manual notify: OK
[x]  * cbFunc hit: 00000002, 00001337, 00000000
[r]   sceKernelDelayThreadCB: 00000000
